Mathlib/Computability/ContextFreeGrammar.lean

```diff
@@ -253,3 +253,172 @@ theorem Language.IsContextFree.reverse (L : Language T) :
   fun ⟨g, hg⟩ => ⟨g.reverse, hg ▸ Set.ext g.mem_reverse_language_iff_reverse_mem_language⟩
 
 end closure_reversal
+
+section embed_project
+
+/-! This section contains only auxiliary constructions that will shorten upcoming proofs of
+closure properties. When combining several grammars together, we usually want to take a sum type of
+their nonterminal types and embed respective nonterminals to this sum type. We subsequently show
+that the resulting grammar preserves derivations of those strings that may contain any terminals but
+only the proper nonterminals. The embedding operation must be injective. The projection operation
+must be injective on those symbols where it is defined. -/
+
+/-- Mapping `Symbol` when it is a nonterminal. -/
+def Symbol.map {N₀ N : Type*} (f : N₀ → N) : Symbol T N₀ → Symbol T N
+  | Symbol.terminal t => Symbol.terminal t
+  | Symbol.nonterminal n => Symbol.nonterminal (f n)
```

Comment:
Why not
```suggestion
/-- Mapping `Symbol` when it is a nonterminal. -/
def Symbol.map {N₀ N T₀ T : Type*} (f : T₀ → T) (g : N₀ → N) : Symbol T₀ N₀ → Symbol T N
  | Symbol.terminal t => Symbol.terminal (f t)
  | Symbol.nonterminal n => Symbol.nonterminal (g n)
```
This matches `Sum.map`

Reply:
Yes but it would introduce additional useless argument.

Reply:
Are you sure you will never need to map the terminal symbols?

Reply:
If it ever happens, I will use a different mapping function that changes terminals only. Throughout the whole file, terminals and nonterminals are treated in "asymmetric" ways. Note that one of them is a type argument and the other is a field of any given grammar.